spec/haskell/src/SEL4/Object/Endpoint.lhs

```diff
@@ -179,14 +184,16 @@ The IPC receive operation is essentially the same as the send operation, but wit
 >                   doIPCTransfer sender (Just epptr) badge canGrant thread
 >                   let call = blockingIPCIsCall senderState
 >                   fault <- threadGet tcbFault sender
->                   case (call, fault, canGrant || canGrantReply, replyOpt) of
->                       (False, Nothing, _, _) -> do
+>                   if (call || isJust fault)
+>                       then if ((canGrant || canGrantReply) && replyOpt /= Nothing)
+>                           then do
+>                               senderSc <- threadGet tcbSchedContext sender
+>                               donate <- return ((senderSc /= Nothing) && not (isJust fault && isTimeoutFault (fromJust fault)))
```

Comment:
Thanks Corey! I agree about changing the function to take `Maybe Fault`, but I'm not sure we know `isJust fault` in this branch. I think that you could have `call = True` but not `isJust fault` because of the `||`

Reply:
The current `isTimeoutFault` has the following nice property
```
isTimeoutFault (fault_map a) = is_timeout_fault a
```
So I'd like to keep it that way if you don't mind.